diff

Allow one to use
|   <textbox type="autocomplete" sizetopopup="none" autocompletepopup="myPopup" />
and
|   <panel type="autocomplete" id="myPopup" width="300"/>
to specify that the popup's width should NOT be set to that of the
textbox. When using this, one must manually set the width on the
popup widget (set to 300 in the example above).

Currently the popup is sized to the textbox's width regardless of the
value of "sizetopopup".

The `.removeAttribute("width")` call was added for this bug:
https://bugzilla.mozilla.org/show_bug.cgi?id=381084

See: https://bugzilla.mozilla.org/show_bug.cgi?id=416250

Index: toolkit/content/widgets/autocomplete.xml
===================================================================
RCS file: /cvsroot/mozilla/toolkit/content/widgets/autocomplete.xml,v
retrieving revision 1.139
diff -U 10 -r1.139 autocomplete.xml
--- toolkit/content/widgets/autocomplete.xml	22 Apr 2008 19:47:44 -0000	1.139
+++ toolkit/content/widgets/autocomplete.xml	27 Aug 2008 23:48:52 -0000
@@ -703,13 +703,15 @@
 
             document.popupNode = null;
 
-            var rect = aElement.getBoundingClientRect();
-            var nav = aElement.ownerDocument.defaultView.QueryInterface(Components.interfaces.nsIInterfaceRequestor)
-                              .getInterface(Components.interfaces.nsIWebNavigation);
-            var docShell = nav.QueryInterface(Components.interfaces.nsIDocShell);
-            var docViewer = docShell.contentViewer.QueryInterface(Components.interfaces.nsIMarkupDocumentViewer);
-            var width = (rect.right - rect.left) * docViewer.fullZoom;
-            this.setAttribute("width", width > 100 ? width : 100);
+            if (aElement.sizetopopup == "always" || aElement.sizetopopup == "pref") {
+              var rect = aElement.getBoundingClientRect();
+              var nav = aElement.ownerDocument.defaultView.QueryInterface(Components.interfaces.nsIInterfaceRequestor)
+                                .getInterface(Components.interfaces.nsIWebNavigation);
+              var docShell = nav.QueryInterface(Components.interfaces.nsIDocShell);
+              var docViewer = docShell.contentViewer.QueryInterface(Components.interfaces.nsIMarkupDocumentViewer);
+              var width = (rect.right - rect.left) * docViewer.fullZoom;
+              this.setAttribute("width", width > 100 ? width : 100);
+            }
 
             // setConsumeRollupEvent() before we call openPopup()
             var nsIPopupBO = Components.interfaces.nsIPopupBoxObject;
@@ -797,7 +799,10 @@
           if (this.mPopupOpen) {
             this.hidePopup();
             document.popupNode = null;
-            this.removeAttribute("width");
+            var sizetopopup = this.mInput.getAttribute("sizetopopup");
+            if (sizetopopup == "always" || sizetopopup == "pref") {
+              this.removeAttribute("width");
+            }
           }
         ]]>
         </body>
@@ -952,10 +957,12 @@
             this.selectedIndex = -1;
             document.popupNode = null;
 
-            var width = aElement.getBoundingClientRect().width;
-            this.setAttribute("width", width > 100 ? width : 100);
-            // invalidate() depends on the width attribute
-            this._invalidate();
+            if (aElement.sizetopopup == "always" || aElement.sizetopopup == "pref") {
+              var width = aElement.getBoundingClientRect().width;
+              this.setAttribute("width", width > 100 ? width : 100);
+              // invalidate() depends on the width attribute
+              this._invalidate();
+            }
 
             // setConsumeRollupEvent() before we call openPopup()
             var nsIPopupBO = Components.interfaces.nsIPopupBoxObject;
